### DISCUSSION TITLE
Interactive CLI / Shell

### DISCUSSION THREAD

**Discussion by vdmitriyev** [dmarro89/dare-db#16] [Ideas]
Database should also be reachable also through a CLI. There should be 2 types of commands for CLI:

* Commands/parameters to start interactive shell/CLI (host, port, login, password, etc.) 
* Commands/parameters to execute in the shell directly in an interactive mode 

#### Parameters for CLI

| **Parameter flag** | **Long flag** | **Description** | **Default** | **Status** | 
|--------------------|---------------|-----------------|-------------|------------|
| -h | --host | Name and port of the server (```<HOST>:PORT```)| localhost:2605 | --- |
| -u | --username | Connect to the database as the user username instead of the default| admin | --- |
| -p | --password | Password for the user | --- | --- |
| -i | --init-config| Creates a config file with default values | --- |
| -c | --config| Shows location of the used config file | --- |

#### Commands for interactive shell

| **Command** |  **Description** |
|-------------|------------------|
| show info | Shows information of the current sessions |
| show schema | Shows information of the current schema |
| password ```<USER>``` | Change user password|
| get```<KEY>``` | Shows value of a key |
| set ```<KEY> <VALUE>```| Set value of a key |

Further materials for inspiration:

- https://www.dolthub.com/blog/2023-03-29-interactive-shell-golang/
- https://appmaster.io/blog/writing-cli-applications-go

**Comment by dmarro89** [comment]
Why use a restart option ?

Reset-admin-user could be a potential security problem as parameter -> every user could reset the credentials and access with default login. I'll not use it as parameter flag. The same as dump-data.
I'd like to consider them as commands.

get-key and set-key -> get - set.

Thanks for your idea!

**Comment by vdmitriyev** [comment]
* ```reset-admin-user``` had been remove from the list
* ```dump-data``` could be useful for the users who runs the database. Database is fully in-memory and there is no possibility to "debug" or make a dump of the current data. The parameter will remove from the list for now in order not to make it CLI more complex.
* ```restart``` is more for developers. E.g., to restart the server after some changes automatically (e.g., "web server" must be restarted everything time a template changes). I would suggest to dump this parameter for now anyway. 

* get-key and set-key -> get - set. -> DONE

**Comment by go-while** [comment]
a socket can be read with textproto and handle commands by "line", no need to split anything.

so we could integrate the cli tool into the main binary and just not boot the dare-db, when called with args.
pass commands to socket or open a interactive shell (local or remote)

i have some socket code to copy and paste *g

**Comment by vdmitriyev** [comment]
Just an idea for future of a possible CLI integration: 

make use of [usql](https://github.com/xo/usql)

**Reply by dmarro89** [reply]
What about Cobra?
It seems easy to use and it's used also by K8s

**Reply by vdmitriyev** [reply]
`cobra` is very handy package to create CLI. Recently I came across a quite small go-project, which helped me to understand technical details, how exactly `cobra` could be used for CLI. Here is the project - https://github.com/arimatakao/mdx

I also found following `cobra` article quite useful: [How to Use Cobra and Viper to Create Your First Golang CLI Tool](https://betterprogramming.pub/step-by-step-using-cobra-and-viper-to-create-your-first-golang-cli-tool-8050d7675093)

Besides `cobra`, `flag` from the standard package could be used. It is quite handy, when a CLI is small and straightforward. See the following article: [Go by Example: Command-Line Flags](https://gobyexample.com/command-line-flags)

**Comment by dmarro89** [comment]
Closing the discussion.
Here the following issue -> https://github.com/dmarro89/dare-db/issues/41
